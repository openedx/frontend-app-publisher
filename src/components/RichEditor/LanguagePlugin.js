import tinymce from 'tinymce/tinymce.min';
import { BROWSER_DEFAULT, languages } from './editorConstants';

tinymce.PluginManager.add('language', (editor) => {
  const replaceText = (lang) => {
    // Depending on the current location of the cursor, replace the current node with
    // the HTML in `lang`
    const newSpanText = lang === BROWSER_DEFAULT ? '<span id="new_span">&#65279</span>' : `<span lang=${lang} id="new_span">&#65279</span>`;
    const selectedNode = editor.selection.getNode();
    if (selectedNode.nodeName === 'SPAN') {
      if (!selectedNode.innerText.trim()) {
        // inner text consists of only whitespace, so replace entire span to reduce nesting
        editor.dom.setOuterHTML(selectedNode, newSpanText);
      } else {
        // inner text does not consist of only whitespace, so append new span to existing HTML
        editor.dom.setOuterHTML(selectedNode, `${selectedNode.outerHTML}${newSpanText}`);
      }
    } else if (selectedNode.nodeName === 'P') {
      // selection is part of a bigger paragraph, so append the new span to existing HTML
      selectedNode.innerHTML += newSpanText;
    } else if (selectedNode.nodeName === 'BR') {
      // text consists of only whitespace, so replace all of it to avoid adding extra whitespace
      editor.dom.setOuterHTML(selectedNode, newSpanText);
    } else {
      // conservatively assume that the selection contains more HTML to the left
      editor.dom.setOuterHTML(selectedNode, `${selectedNode.outerHTML}${newSpanText}`);
    }
    const newSpan = editor.dom.get('new_span');
    editor.selection.select(newSpan);
    editor.selection.collapse(false); // moves cursor to end of selection
    newSpan.removeAttribute('id');
  };

  const openDialog = (buttonApi) => {
    if (!editor.selection.isCollapsed()) {
      editor.notificationManager.open({
        text: 'You cannot set the language of a selection. Change the language first and then continue typing.',
        type: 'error',
      });
    } else {
      const selectedNode = editor.selection.getNode();
      const currentLang = selectedNode.lang ? selectedNode.lang : BROWSER_DEFAULT;
      editor.windowManager.open({
        title: 'Language plugin',
        body: {
          type: 'panel',
          items: [
            {
              type: 'htmlpanel',
              html: `<div>Current language: ${languages[currentLang].nativeName}</div>`,
            },
            {
              type: 'selectbox',
              name: 'language',
              label: 'Language',
              items: Object.keys(languages).map(lang => ({
                value: lang,
                text: languages[lang].nativeName,
              })),
            },
          ],
        },
        buttons: [
          {
            type: 'cancel',
            text: 'Close',
          },
          {
            type: 'submit',
            text: 'Save',
            primary: true,
          },
        ],
        onSubmit(api) {
          const data = api.getData();
          editor.focus();
          editor.undoManager.transact(() => {
            replaceText(data.language);
            buttonApi.setActive(data.language !== BROWSER_DEFAULT);
          });
          api.close();
        },
      });
    }
  };

  editor.ui.registry.addToggleButton('language', {
    text: 'Foreign language',
    onAction(buttonApi) {
      openDialog(buttonApi);
    },
    onSetup(buttonApi) {
      editor.addShortcut('Meta+L', 'Switch to default language', () => {
        const selectedNode = editor.selection.getNode();
        const currentLang = selectedNode.lang ? selectedNode.lang : BROWSER_DEFAULT;
        if (currentLang !== BROWSER_DEFAULT) {
          editor.undoManager.transact(() => {
            replaceText(BROWSER_DEFAULT);
            buttonApi.setActive(false);
          });
        }
      });
      const updateCurrentLanguage = () => {
        if (editor.selection.getNode().lang) buttonApi.setActive(true);
        else buttonApi.setActive(false);
      };
      editor.on('keyup', updateCurrentLanguage);
      editor.on('click', updateCurrentLanguage);
      return () => {
        // remove event listeners on teardown
        editor.off('keyup', updateCurrentLanguage);
        editor.off('click', updateCurrentLanguage);
      };
    },
  });

  return {
    getMetadata() {
      return {
        name: 'Language plugin',
      };
    },
  };
});
