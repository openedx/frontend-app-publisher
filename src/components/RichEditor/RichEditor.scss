.disabled-rich-text .tox .tox-edit-area__iframe {
  background-color: var(--pgn-color-form-input-bg-disabled);
}
