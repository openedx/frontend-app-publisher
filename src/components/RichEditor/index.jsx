import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import { Editor } from '@tinymce/tinymce-react';
import 'tinymce/tinymce.min';
import 'tinymce/icons/default';
import 'tinymce/plugins/legacyoutput';
import 'tinymce/plugins/link';
import 'tinymce/plugins/lists';
import 'tinymce/plugins/directionality';
import 'tinymce/themes/silver/theme';
import '@edx/tinymce-language-selector';
import 'tinymce/skins/ui/oxide/skin.css';
import contentCss from 'tinymce/skins/content/default/content.min.css';
import contentUiCss from 'tinymce/skins/ui/oxide/content.min.css';
import { Alert } from '@edx/paragon';

class RichEditor extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      charCount: 0,
    };
    this.initCharCount = this.initCharCount.bind(this);
    this.updateCharCount = this.updateCharCount.bind(this);
  }

  initCharCount(event, editor) {
    const content = editor.getContent({ format: 'text' });
    this.setState({
      charCount: content.length,
    });
  }

  updateCharCount(event, editor) {
    const content = editor.getContent({ format: 'text' });
    this.setState({
      charCount: content.length,
    });
    const htmlContent = editor.getContent();
    this.props.input.onChange(htmlContent);
  }

  render() {
    const {
      maxChars,
      id,
      label,
      input: {
        value,
        name,
      },
      meta: {
        submitFailed,
        error,
      },
      disabled,
    } = this.props;
    const remainingChars = maxChars - this.state.charCount;
    const characterLimitMessage = `Recommended character limit (including spaces) is
      ${maxChars}. ${remainingChars} characters remaining.`;

    let contentStyle;
    // In the test environment this causes an error so set styles to empty since they aren't needed for testing.
    try {
      contentStyle = [contentCss, contentUiCss].join('\n');
    } catch (err) {
      contentStyle = '';
    }

    return (
      <div className="form-group">
        <div id={id} name={name} tabIndex="-1" className="mb-2">{label}</div>
        {submitFailed && error
          && (
          <Alert variant="danger">{error}</Alert>
          )}
        {/*
          We are using aria-labelledby here instead of a <label> tag because the TinyMCE Editor
          adds an iframe instead of a standard tag that would be used with <label>. This was
          chosen as the best way to still label what the editor is used for.
          The <div> with the title stating what the Editor is for has the corresponding
          id to match the aria-labelledby.
        */}
        <div aria-labelledby={id} className={classNames({ 'disabled-rich-text': disabled })}>
          <Editor
            initialValue={value}
            init={{
              branding: false,
              menubar: false,
              plugins: 'legacyoutput link lists language directionality',
              statusbar: false,
              selector: `#${id}`,
              toolbar: 'undo redo | bold italic underline | bullist numlist | link | language | ltr rtl',
              entity_encoding: 'raw',
              extended_valid_elements: 'span[lang|id] -span',
              content_css: false,
              content_style: contentStyle,
              default_link_target: '_blank',
            }}
            onChange={this.updateCharCount}
            onKeyUp={this.updateCharCount}
            onInit={this.initCharCount}
            disabled={disabled}
          />
          {maxChars && <span>{characterLimitMessage}</span>}
        </div>
      </div>
    );
  }
}

RichEditor.defaultProps = {
  maxChars: null,
  disabled: false,
};

RichEditor.propTypes = {
  id: PropTypes.string.isRequired,
  label: PropTypes.oneOfType([PropTypes.string, PropTypes.element]).isRequired,
  maxChars: PropTypes.number,
  input: PropTypes.shape({
    value: PropTypes.string,
    name: PropTypes.string,
    onChange: PropTypes.func.isRequired,
  }).isRequired,
  meta: PropTypes.shape({
    touched: PropTypes.bool,
    submitFailed: PropTypes.bool,
    error: PropTypes.string,
  }).isRequired,
  disabled: PropTypes.bool,
};

export default RichEditor;
