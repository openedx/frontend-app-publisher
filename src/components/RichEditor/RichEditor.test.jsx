import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';

import RichEditor from './index';

describe('RichEditor', () => {
  it('shows a rich text editor with no default text value', () => {
    const component = shallow(<RichEditor
      label="Rich Text Editor Test"
      id="rich-text-editor-test"
      input={{
        value: null,
        onChange: () => null,
      }}
      maxChars={500}
      meta={{ touched: false, error: '' }}
    />);
    expect(shallowToJson(component)).toMatchSnapshot();
  });

  it('shows a rich text editor with default text value', () => {
    const component = shallow(<RichEditor
      label={<strong>Rich Text Editor Test</strong>}
      id="rich-text-editor-test"
      input={{
        value: '<p>Prior text<p>',
        onChange: () => null,
      }}
      maxChars={2500}
      meta={{ touched: false, error: '' }}
    />);
    expect(shallowToJson(component)).toMatchSnapshot();
  });

  it('shows a rich text editor with no maxChars', () => {
    const component = shallow(<RichEditor
      label={<strong>Rich Text Editor Test</strong>}
      id="rich-text-editor-test"
      input={{
        value: '<p>Prior text<p>',
        onChange: () => null,
      }}
      meta={{ touched: false, error: '' }}
    />);
    expect(shallowToJson(component)).toMatchSnapshot();
  });

  it('shows a rich text editor and an error', () => {
    const component = shallow(<RichEditor
      label="Rich Text Editor Test"
      id="rich-text-editor-test"
      input={{
        value: null,
        onChange: () => null,
      }}
      maxChars={500}
      meta={{ touched: true, error: 'Required' }}
    />);
    expect(shallowToJson(component)).toMatchSnapshot();
  });
});
