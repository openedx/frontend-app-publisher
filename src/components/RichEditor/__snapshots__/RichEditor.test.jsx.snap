// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RichEditor shows a rich text editor and an error 1`] = `
<div
  className="form-group"
>
  <div
    className="mb-2"
    id="rich-text-editor-test"
    tabIndex="-1"
  >
    Rich Text Editor Test
  </div>
  <div
    aria-labelledby="rich-text-editor-test"
    className=""
  >
    <Editor
      cloudChannel="5"
      disabled={false}
      init={
        Object {
          "branding": false,
          "content_css": false,
          "content_style": "",
          "default_link_target": "_blank",
          "entity_encoding": "raw",
          "extended_valid_elements": "span[lang|id] -span",
          "menubar": false,
          "plugins": "legacyoutput link lists language directionality",
          "selector": "#rich-text-editor-test",
          "statusbar": false,
          "toolbar": "undo redo | bold italic underline | bullist numlist | link | language | ltr rtl",
        }
      }
      initialValue={null}
      onChange={[Function]}
      onInit={[Function]}
      onKeyUp={[Function]}
    />
    <span>
      Recommended character limit (including spaces) is
      500. 500 characters remaining.
    </span>
  </div>
</div>
`;

exports[`RichEditor shows a rich text editor with default text value 1`] = `
<div
  className="form-group"
>
  <div
    className="mb-2"
    id="rich-text-editor-test"
    tabIndex="-1"
  >
    <strong>
      Rich Text Editor Test
    </strong>
  </div>
  <div
    aria-labelledby="rich-text-editor-test"
    className=""
  >
    <Editor
      cloudChannel="5"
      disabled={false}
      init={
        Object {
          "branding": false,
          "content_css": false,
          "content_style": "",
          "default_link_target": "_blank",
          "entity_encoding": "raw",
          "extended_valid_elements": "span[lang|id] -span",
          "menubar": false,
          "plugins": "legacyoutput link lists language directionality",
          "selector": "#rich-text-editor-test",
          "statusbar": false,
          "toolbar": "undo redo | bold italic underline | bullist numlist | link | language | ltr rtl",
        }
      }
      initialValue="<p>Prior text<p>"
      onChange={[Function]}
      onInit={[Function]}
      onKeyUp={[Function]}
    />
    <span>
      Recommended character limit (including spaces) is
      2500. 2500 characters remaining.
    </span>
  </div>
</div>
`;

exports[`RichEditor shows a rich text editor with no default text value 1`] = `
<div
  className="form-group"
>
  <div
    className="mb-2"
    id="rich-text-editor-test"
    tabIndex="-1"
  >
    Rich Text Editor Test
  </div>
  <div
    aria-labelledby="rich-text-editor-test"
    className=""
  >
    <Editor
      cloudChannel="5"
      disabled={false}
      init={
        Object {
          "branding": false,
          "content_css": false,
          "content_style": "",
          "default_link_target": "_blank",
          "entity_encoding": "raw",
          "extended_valid_elements": "span[lang|id] -span",
          "menubar": false,
          "plugins": "legacyoutput link lists language directionality",
          "selector": "#rich-text-editor-test",
          "statusbar": false,
          "toolbar": "undo redo | bold italic underline | bullist numlist | link | language | ltr rtl",
        }
      }
      initialValue={null}
      onChange={[Function]}
      onInit={[Function]}
      onKeyUp={[Function]}
    />
    <span>
      Recommended character limit (including spaces) is
      500. 500 characters remaining.
    </span>
  </div>
</div>
`;

exports[`RichEditor shows a rich text editor with no maxChars 1`] = `
<div
  className="form-group"
>
  <div
    className="mb-2"
    id="rich-text-editor-test"
    tabIndex="-1"
  >
    <strong>
      Rich Text Editor Test
    </strong>
  </div>
  <div
    aria-labelledby="rich-text-editor-test"
    className=""
  >
    <Editor
      cloudChannel="5"
      disabled={false}
      init={
        Object {
          "branding": false,
          "content_css": false,
          "content_style": "",
          "default_link_target": "_blank",
          "entity_encoding": "raw",
          "extended_valid_elements": "span[lang|id] -span",
          "menubar": false,
          "plugins": "legacyoutput link lists language directionality",
          "selector": "#rich-text-editor-test",
          "statusbar": false,
          "toolbar": "undo redo | bold italic underline | bullist numlist | link | language | ltr rtl",
        }
      }
      initialValue="<p>Prior text<p>"
      onChange={[Function]}
      onInit={[Function]}
      onKeyUp={[Function]}
    />
  </div>
</div>
`;
